src/adapters/tcp.rs

```diff
@@ -78,7 +97,49 @@ impl Remote for RemoteResource {
             _ => panic!("Internal error: Got wrong config"),
         };
         let peer_addr = *remote_addr.socket_addr();
-        let stream = TcpStream::connect(peer_addr)?;
+
+        let socket = Socket::new(
+            match peer_addr {
+                SocketAddr::V4 { .. } => Domain::IPV4,
+                SocketAddr::V6 { .. } => Domain::IPV6,
+            },
+            Type::STREAM,
+            Some(Protocol::TCP),
+        )?;
+        socket.set_nonblocking(true)?;
+
+        if let Some(source_address) = config.source_address {
+            socket.bind(&source_address.into())?;
+        }
+
+        #[cfg(unix)]
+        if let Some(bind_device) = config.bind_device {
+            let device = CString::new(bind_device)?;
+
+            #[cfg(not(target_os = "macos"))]
+            socket.bind_device(Some(device.as_bytes()))?;
+
+            #[cfg(target_os = "macos")]
+            match NonZeroU32::new(unsafe { libc::if_nametoindex(device.as_ptr()) }) {
+                Some(index) => socket.bind_device_by_index(Some(index))?,
+                None => {
+                    return Err(io::Error::new(
+                        ErrorKind::NotFound,
+                        "Bind device interface not found",
+                    ))
+                }
+            }
+        }
+
+        match socket.connect(&peer_addr.into()) {
+            #[cfg(unix)]
+            Err(e) if e.raw_os_error() != Some(libc::EINPROGRESS) => return Err(e),
+            #[cfg(windows)]
+            Err(e) if e.kind() != io::ErrorKind::WouldBlock => return Err(e),
```

Comment:
Ask from the unknowledge, could a `WouldBlock` be obtained on Unix at this point (it seems like it's totally fine returning that for a socket implementation at this point)? In this case, an error will be returned when we shouldn't

Reply:
I copied that bit from mio:
- Linux: https://github.com/tokio-rs/mio/blob/5234b5f19dd79b817cd74fc4fe25c9aad2e3e24c/src/sys/unix/tcp.rs#L31
- Windows: https://github.com/tokio-rs/mio/blob/5234b5f19dd79b817cd74fc4fe25c9aad2e3e24c/src/sys/windows/tcp.rs#L49

Reply:
Then it should be ok 👍🏻